gprofiler/system_metrics.py

```diff
@@ -0,0 +1,91 @@
+import statistics
+import time
+from collections import deque
+from threading import Event, Thread
+from typing import Deque, List, Optional, Tuple
+
+import psutil
+
+from gprofiler.exceptions import ThreadStopTimeoutError
+
+DEFAULT_POLLING_INTERVAL_SECONDS = 5
+STOP_TIMEOUT_SECONDS = 30
+
+
+class SystemMetricsMonitor:
+    def __init__(self, max_memory_poll_age_seconds: int, polling_rate_seconds: int = DEFAULT_POLLING_INTERVAL_SECONDS):
+        self._polling_rate_seconds = polling_rate_seconds
+        self._cpu_count = psutil.cpu_count() or 1
+        self._mem_percentages: Deque[float] = deque(maxlen=round(max_memory_poll_age_seconds / polling_rate_seconds))
+        self._last_cpu_poll_time: Optional[float] = None
+        self._last_cpu_utilization_percentages: Optional[Tuple[float, float]] = None
+        self._stop_event = Event()
+        self._thread = None
+
+        self.get_cpu_utilization()  # Call this once to set the necessary data
+
+    def start(self):
+        assert self._thread is None, "SystemMetricsMonitor is already running"
+        self._stop_event.clear()
+        self._thread = Thread(target=self._continuously_polling_memory, args=(self._polling_rate_seconds,))
+        self._thread.start()
+
+    def stop(self):
+        self._stop_event.set()
+        assert self._thread is not None, "SystemMetricsMonitor was not running"
+        self._thread.join(STOP_TIMEOUT_SECONDS)
+        if self._thread.is_alive():
+            raise ThreadStopTimeoutError("The SystemMetricsMonitor internal thread could not stop because of a timeout")
+        self._thread = None
+
+    def _continuously_polling_memory(self, polling_rate_seconds: int):
+        while not self._stop_event.is_set():
+            start_time = time.monotonic()
+            current_ram_percent = psutil.virtual_memory().percent
+            self._mem_percentages.append(current_ram_percent)
+            elapsed = time.monotonic() - start_time
+            self._sleep(polling_rate_seconds - elapsed)
+
+    def _sleep(self, amount_seconds: float):
+        sleep_intervals_total = int(amount_seconds * 100)
+        for sleep_interval in range(sleep_intervals_total):
+            time.sleep(sleep_interval / 100)
+            if self._stop_event.is_set():
+                return
+
+    def get_average_memory_utilization(self) -> Optional[float]:
+        current_length = len(self._mem_percentages)
+        if current_length == 0:
+            return None
+        all_percentages: List[float] = []
+        # Avoid a race condition by only fetching the current X values. More values can be added in the meantime by
```

Comment:
What race? Why is it racy? It's okay, just pop all values until none are left

Reply:
This can also just be a list and you can swap it with an empty one before calculating the mean. Will simplify the code